compiler/rustc_mir_transform/src/coroutine/by_move_body.rs

```diff
@@ -1,7 +1,63 @@
-//! A MIR pass which duplicates a coroutine's body and removes any derefs which
-//! would be present for upvars that are taken by-ref. The result of which will
-//! be a coroutine body that takes all of its upvars by-move, and which we stash
-//! into the `CoroutineInfo` for all coroutines returned by coroutine-closures.
+//! This pass constructs a second closure body sufficient for `FnOnce`/`AsyncFnOnce`
```

Comment:
This is high-level motivation for this pass -- it's probably not as detailed as it could be, but it's a start.